core/src/main/scala/org/apache/spark/status/AppStatusStore.scala

```diff
@@ -35,8 +36,24 @@ private[spark] class AppStatusStore(
     val store: KVStore,
     val listener: Option[AppStatusListener] = None) {
 
+  /**
+   * This method contains an automatic retry logic and tries to get a valid [[v1.ApplicationInfo]].
+   * See [SPARK-31632] The ApplicationInfo in KVStore may be accessed before it's prepared
+   */
   def applicationInfo(): v1.ApplicationInfo = {
```

Comment:
Instead of make this function blocking, how about capture the Exception outside and display the proper message?

Reply:
If blocking is supposed to be harmful, I'd prefer an optional result instead of throwing an exception here. In fact, due to the previous design and implementation, all the invokers of the method are expecting a valid `ApplicationInfo`, which means they are not capable of dealing with a `None` value or an exception now. If we change the returned type to `Option`, all the logic of the invokers (and invokers of these invokers maybe) should be somehow adjusted. There will be more design choices then.

Adding the retry logic solves the problem from the root. As I said, however, I can definitely make another choice if we reach a consensus.

Reply:
I think capturing the exception outside is actually a good compromise considering the issue is rather a corner case. Let's scope narrow here since the issue is rather minor, and consider a better fix with a standard approach when any bigger issue is found in the design.

Reply:
Also sounds reasonable to me.

Reply:
Thanks for your feedback, @HyukjinKwon. However, I didn't get your point (sorry for that). Did you mean changing the implementation to capturing the exception outside, or let's just keep the current approach?

Reply:
Sorry I wasn't clear. I was thinking a way, for example, as below:

```diff
-    store.view(classOf[ApplicationInfoWrapper]).max(1).iterator().next().info
+    try {
+      store.view(classOf[ApplicationInfoWrapper]).max(1).iterator().next().info
+    } catch {
+      case _: NoSuchElementException =>
+        throw new NoSuchElementException(
+          "Application information was not found. If your Spark application is " +
+            "starting up, try again when your Spark application is ready.")
+    }
```

Reply:
Thanks for the explanation. I'll update the PR as suggested.